static/admin/components.js

```diff
@@ -0,0 +1,109 @@
+CMS.registerEditorComponent({
+    id: "image-block",
+    label: "Image Block",
+    fields: [
+        { name: "image", label: "Image", widget: "image",  default: "" },
+    ],
+    pattern: /{{< image-block image="(.*)" >}}/,
+    fromBlock: function(match) {
+        return {
+            "image": match[1],
+        };
+    },
+    toBlock: function(obj) {
+        return `{{< image-block image="${obj["image"]}" >}}`;
+    },
+    toPreview: function(obj) {
+        return `{{< image-block image="${obj["image"]}" >}}`;
+    },
+});
+
+CMS.registerEditorComponent({
+    id: "text-block",
+    label: "Text Block",
+    fields: [
+        { name: "title", label: "Title", widget: "string", default: "" },
+        { name: "text", label: "Text", widget: "string", default: "" },
+        { name: "button-text-1", label: "Button Text 1", widget: "string", default: "" },
+        { name: "button-url-1", label: "Button URL 1", widget: "string", default: "" },
+        { name: "button-text-2", label: "Button Text 2", widget: "string", default: "" },
+        { name: "button-url-2", label: "Button URL 2", widget: "string", default: "" },
+    ],
+    pattern: /{{< text-block title="(.*)" text="(.*)" button-text-1="(.*)" button-url-1="(.*)" button-text-2="(.*)" button-url-2="(.*)" >}}/,
+    fromBlock: function(match) {
+        return {
+            "title": match[1],
+            "text": match[2],
+            "button-text-1": match[3],
+            "button-url-1": match[4],
+            "button-text-2": match[5],
+            "button-url-2": match[6],
+        };
+    },
+    toBlock: function(obj) {
+        return `{{< text-block title="${obj["title"]}" text="${obj["text"]}" button-text-1="${obj["button-text-1"]}" button-url-1="${obj["button-url-1"]}" button-text-2="${obj["button-text-2"]}" button-url-2="${obj["button-url-2"]}" >}}`;
+    },
+    toPreview: function(obj) {
```

Comment:
Ignore this if it's not the convention: Is there anyway to re-use these functions? It seems like the `toBlock` and `toPreview` methods are the exact same.

Reply:
yeah its a bit weird, most sources i've seen them as copies of one another, we could have a different output for toPreview if we wanted to get the cms preview thing working properly